File: packages/var-rpc/test/subscribe-sync.test.ts
import { afterEach, describe, expect, it } from 'vitest';

import {
  VarManager,
  type DocumentSchemaDefinition,
  type NormalizedVarSourceDefinition,
  type ProjectedVarSourceDefinition,
  type VarPushEvent,
  type VarSourceProvider,
  type VarSourceProviderContext,
} from '@kitsy/cnos-core';
import { createVarEngine, memoryStore, type VarAuthorize, type VarEngine, type VarStore } from '@kitsy/cnos-var-server';

import { createRpcVarProvider, serveVarRpc, type RunningVarRpcServer } from '../src/index.js';

/**
 * SELF-SYNCHRONIZING SUBSCRIBE (round-3 follow-up).
 *
 * Round-3 blocker 1 ("rpc reconnect never re-pulls subscribed scopes") was fixed on the CLIENT
 * side: every (re)connect reports through `onSubscriptionConnected` and the SDK re-pulls. That
 * left a server-side hole with the SAME failure mode: `attachVarRpc` registered its
 * `engine.onCommit` listener only AFTER `await authorize(...)`, so a commit landing between the
 * Subscribe request arriving and that registration completing was delivered by NEITHER the
 * stream (no hook yet) NOR the resync pull (possibly issued just before the commit). The window
 * is sub-millisecond, and losing a DEACTIVATION in it leaves a consumer serving withdrawn policy
 * with no poller to converge — subscribe-capable sources deliberately do not poll.
 *
 * The stream is now self-synchronizing: the listener is registered synchronously and buffers
 * across the authorization window, and an accepted Subscribe emits the current state as its
 * first event. These tests drive the race DETERMINISTICALLY (authorize blocks on a released
 * promise) rather than hoping to hit a sub-millisecond window, and they neutralize the client
 * resync pull where the point is that the STREAM ALONE converges.
 */

const AGENTIC_SCHEMA: DocumentSchemaDefinition = {
  fields: {
    enabled: { type: 'boolean', required: true },
    model_target_ref: { type: 'string', required: true },
  },
  additionalProperties: false,
};

const documents = { 'agentic-lanes/v1': AGENTIC_SCHEMA };
const delay = (ms: number): Promise<void> => new Promise((resolve) => setTimeout(resolve, ms));

async function until(predicate: () => boolean, timeoutMs = 8000, stepMs = 10): Promise<boolean> {
  const deadline = Date.now() + timeoutMs;
  while (!predicate() && Date.now() < deadline) {
    await delay(stepMs);
  }
  return predicate();
}

const servers: RunningVarRpcServer[] = [];
const providers: VarSourceProvider[] = [];

function track<T extends VarSourceProvider>(provider: T): T {
  providers.push(provider);
  return provider;
}

afterEach(async () => {
  await Promise.all(providers.splice(0).map((provider) => provider.close().catch(() => undefined)));
  await Promise.all(servers.splice(0).map((server) => server.close().catch(() => undefined)));
});

async function activate(engine: VarEngine, scope: string, document: unknown, expectedGeneration: number): Promise<void> {
  const created = await engine.createRevision({
    scope,
    document,
    ...(scope.includes('.') ? { schemaId: 'agentic-lanes/v1' } : {}),
  });
  await engine.activate({ scope, revision: created.revision, expectedGeneration });
}

/**
 * An `authorize` hook that BLOCKS until released. `entered` counts calls that have reached the
 * await, which is the precise moment the server is inside the authorization window with its
 * commit listener already registered — so a commit issued after `entered` becomes 1 lands
 * exactly in the window under test. No sleeps, no sub-millisecond guessing.
 */
function gatedAuthorize(verdict: boolean): { authorize: VarAuthorize; entered: () => number; release: () => void } {
  let entered = 0;
  let release!: () => void;
  const gate = new Promise<void>((resolve) => {
    release = resolve;
  });

  return {
    authorize: async () => {
      entered += 1;
      await gate;
      return verdict;
    },
    entered: () => entered,
    release,
  };
}

async function harness(authorize?: VarAuthorize): Promise<{
  store: VarStore;
  engine: VarEngine;
  server: RunningVarRpcServer;
}> {
  const store = memoryStore();
  const engine = createVarEngine(store, { documents });
  const server = await serveVarRpc(store, { engine, documents, ...(authorize ? { authorize } : {}) });
  servers.push(server);
  return { store, engine, server };
}

function providerFor(target: string): VarSourceProvider {
  const def: ProjectedVarSourceDefinition = { transport: 'rpc', url: target, auth: {} };
  return track(createRpcVarProvider(def, { resolveSecret: async () => '' }));
}

// ---------------------------------------------------------------------------
// 1. The authorization window
// ---------------------------------------------------------------------------

describe('a commit landing in the authorize window', () => {
  it('is BUFFERED and delivered exactly once once authorization succeeds', async () => {
    const gate = gatedAuthorize(true);
    const { engine, server } = await harness(gate.authorize);
    const provider = providerFor(server.target);

    const events: VarPushEvent[] = [];
    const stop = provider.subscribe?.([{ group: 'agentic' }], (event) => events.push(event));

    // The server is now inside `await authorize(...)`. Before the fix, its commit listener did
    // not exist yet and this commit was dropped on the floor.
    expect(await until(() => gate.entered() === 1)).toBe(true);
    await activate(engine, 'agentic', { 'agentic.lanes.vinci': { enabled: true, model_target_ref: 'in-window' } }, 0);
    expect(events).toHaveLength(0); // nothing is written while authorization is pending

    gate.release();

    expect(await until(() => events.length > 0)).toBe(true);
    expect(events[0]?.kind).toBe('batch');
    expect(events[0]?.batch?.values).toEqual({
      'agentic.lanes.vinci': { enabled: true, model_target_ref: 'in-window' },
    });

    // EXACTLY once: the flushed buffer and the initial state carry the same revision, and the
    // initial state is deduplicated against the flush rather than repeated.
    await delay(400);
    expect(events).toHaveLength(1);

    stop?.();
  }, 30_000);

  it('is DISCARDED, unwritten, when authorization fails — and the stream still terminates', async () => {
    const gate = gatedAuthorize(false);
    const { engine, server } = await harness(gate.authorize);

    const failures: { terminal: boolean }[] = [];
    const def: ProjectedVarSourceDefinition = { transport: 'rpc', url: server.target, auth: {} };
    const provider = track(
      createRpcVarProvider(def, { resolveSecret: async () => '' }, {
        onError: (_error, info) => failures.push({ terminal: info.terminal }),
      }),
    );

    const events: VarPushEvent[] = [];
    const stop = provider.subscribe?.([{ group: 'agentic' }], (event) => events.push(event));

    expect(await until(() => gate.entered() === 1)).toBe(true);
    await activate(engine, 'agentic', { 'agentic.lanes.vinci': { enabled: true, model_target_ref: 'refused' } }, 0);

    gate.release();

    // Terminal, per the unchanged UNAUTHENTICATED policy — not weakened by the buffering.
    expect(await until(() => failures.length > 0)).toBe(true);
    expect(failures[0]?.terminal).toBe(true);

    // And the buffered commit was discarded, never written to a refused identity.
    await delay(400);
    expect(events).toHaveLength(0);

    stop?.();
  }, 30_000);
});

// ---------------------------------------------------------------------------
// 2. The initial event
// ---------------------------------------------------------------------------

describe('an accepted Subscribe emits the current state first', () => {
  it('a fresh subscribe immediately receives the current head, with no commit at all', async () => {
    const { engine, server } = await harness();
    await activate(engine, 'agentic', { 'agentic.lanes.vinci': { enabled: true, model_target_ref: 'already-live' } }, 0);

    const provider = providerFor(server.target);
    const events: VarPushEvent[] = [];
    const stop = provider.subscribe?.([{ group: 'agentic' }], (event) => events.push(event));

    // Nothing is committed after the subscribe: this can only be the initial state event.
    expect(await until(() => events.length > 0)).toBe(true);
    expect(events[0]?.kind).toBe('batch');
    expect(events[0]?.batch?.values).toEqual({
      'agentic.lanes.vinci': { enabled: true, model_target_ref: 'already-live' },
    });
    expect(events[0]?.batch?.generation).toBe(1);

    await delay(300);
    expect(events).toHaveLength(1);
    stop?.();
  }, 30_000);

  it('a subscribe to a DEACTIVATED scope immediately receives no_head', async () => {
    const { engine, server } = await harness();
    await activate(engine, 'agentic', { 'agentic.lanes.vinci': { enabled: true, model_target_ref: 'r' } }, 0);
    await engine.deactivate({ scope: 'agentic', expectedGeneration: 1 });

    const provider = providerFor(server.target);
    const events: VarPushEvent[] = [];
    const stop = provider.subscribe?.([{ group: 'agentic' }], (event) => events.push(event));

    expect(await until(() => events.length > 0)).toBe(true);
    expect(events[0]).toEqual({ kind: 'no-head', scope: 'agentic' });
    stop?.();
  }, 30_000);

  it('a subscribe to a never-activated scope immediately receives no_head', async () => {
    const { server } = await harness();
    const provider = providerFor(server.target);
    const events: VarPushEvent[] = [];
    const stop = provider.subscribe?.([{ group: 'agentic' }], (event) => events.push(event));

    expect(await until(() => events.length > 0)).toBe(true);
    expect(events[0]).toEqual({ kind: 'no-head', scope: 'agentic' });
    stop?.();
  }, 30_000);
});

// ---------------------------------------------------------------------------
// 3. Convergence from the STREAM ALONE (client resync pull neutralized)
// ---------------------------------------------------------------------------

const staticDocument = { enabled: false, model_target_ref: 'static-tier' };

/**
 * A manager whose rpc provider is built with `onSubscriptionConnected` STRIPPED from its
 * context, so the client-side reconnect resync pull of `d50f34a` never fires. Whatever
 * convergence these tests observe therefore came from the subscription stream alone.
 */
async function streamOnlyHarness(): Promise<{
  engine: VarEngine;
  manager: VarManager;
  resyncPulls: () => number;
  down: () => Promise<void>;
  restart: () => Promise<void>;
}> {
  const store = memoryStore();
  const engine = createVarEngine(store, { documents });
  let server = await serveVarRpc(store, { engine, documents, port: 0 });
  const port = server.port;
  servers.push(server);

  let resyncPulls = 0;

  const source: NormalizedVarSourceDefinition = {
    transport: 'rpc',
    url: server.target,
    auth: {},
    // Declared deliberately: a subscribe-capable source IGNORES pollInterval (capability rule),
    // so convergence below cannot have come from a poller either.
    pollInterval: '20ms',
  };

  const manager = new VarManager({
    varSources: { ops: source },
    vars: { agentic: { source: 'ops', mode: 'prefetch' } },
    documents,
    schema: { 'var.agentic.lanes.vinci': { document: 'agentic-lanes/v1' } },
    providerModules: [
      {
        transport: 'rpc',
        create: (def, ctx) => {
          // Strip the resync seam: count invocations so the test can PROVE it was neutralized
          // rather than merely assuming so.
          const { onSubscriptionConnected: _stripped, ...rest } = ctx as VarSourceProviderContext & {
            onSubscriptionConnected?: unknown;
          };
          void _stripped;
          return track(
            createRpcVarProvider(def, {
              ...(rest as VarSourceProviderContext),
              onSubscriptionConnected: () => {
                resyncPulls += 1;
              },
            }),
          );
        },
      },
    ],
    resolveSecret: async () => '',
    warn: () => undefined,
  });

  manager.setOverlayReader((key) => (key === 'var.agentic.lanes.vinci' ? staticDocument : undefined));
  manager.setFallbackSnapshotReader((key) =>
    key === 'var.agentic.lanes.vinci'
      ? { value: staticDocument, source: 'static', freshness: 'fresh' }
      : undefined,
  );

  return {
    engine,
    manager,
    resyncPulls: () => resyncPulls,
    down: async () => {
      await server.close();
      servers.splice(servers.indexOf(server), 1);
    },
    restart: async () => {
      server = await serveVarRpc(store, { engine, documents, port });
      servers.push(server);
    },
  };
}

describe('reconnect converges from the STREAM ALONE (resync pull neutralized)', () => {
  it('an ACTIVATION missed during the outage converges', async () => {
    const { engine, manager, resyncPulls, down, restart } = await streamOnlyHarness();

    await activate(engine, 'agentic', { 'agentic.lanes.vinci': { enabled: true, model_target_ref: 'before' } }, 0);
    await manager.start();
    expect(manager.readRuntimeVar('var.agentic.lanes.vinci')).toEqual({
      enabled: true,
      model_target_ref: 'before',
    });

    await delay(200);
    await down();

    // EXACTLY ONE mutation, entirely while the client is disconnected.
    await activate(engine, 'agentic', { 'agentic.lanes.vinci': { enabled: true, model_target_ref: 'during-outage' } }, 1);

    await restart();

    expect(
      await until(
        () =>
          (manager.readRuntimeVar('var.agentic.lanes.vinci') as { model_target_ref?: string } | undefined)
            ?.model_target_ref === 'during-outage',
        15_000,
      ),
    ).toBe(true);

    // The resync seam was replaced by a counter that pulls nothing: the reconnect DID happen,
    // and no pull was issued on its behalf.
    expect(resyncPulls()).toBeGreaterThan(0);

    await manager.close();
  }, 40_000);

  it('a DEACTIVATION missed during the outage converges and falls back to the static tier', async () => {
    const { engine, manager, resyncPulls, down, restart } = await streamOnlyHarness();

    await activate(engine, 'agentic', { 'agentic.lanes.vinci': { enabled: true, model_target_ref: 'before' } }, 0);
    await manager.start();
    expect(manager.readRuntimeVar('var.agentic.lanes.vinci')).toEqual({
      enabled: true,
      model_target_ref: 'before',
    });

    const observed: { source: string; value: unknown }[] = [];
    manager.watch('var.agentic.lanes.vinci', (next) => observed.push({ source: next.source, value: next.value }));

    await delay(200);
    await down();

    // The worst case, and the one with NO other path to convergence: a withdrawal the client
    // never hears about, on a source that runs no poller. EXACTLY ONE mutation.
    await engine.deactivate({ scope: 'agentic', expectedGeneration: 1 });

    await restart();

    expect(await until(() => manager.readRuntimeVar('var.agentic.lanes.vinci') === undefined, 15_000)).toBe(true);
    expect(manager.status()['agentic.lanes.vinci']?.source).toBe('static');
    expect(observed).toEqual([{ source: 'static', value: staticDocument }]);
    expect(resyncPulls()).toBeGreaterThan(0);

    await manager.close();
  }, 40_000);

  it('does NOT fire a watcher when the initial event repeats the revision the client already has', async () => {
    const { engine, manager, down, restart } = await streamOnlyHarness();

    await activate(engine, 'agentic', { 'agentic.lanes.vinci': { enabled: true, model_target_ref: 'stable' } }, 0);
    await manager.start();

    const fires: unknown[] = [];
    manager.watch('var.agentic.lanes.vinci', (next) => fires.push(next.value));

    await delay(200);
    await down();
    // NOTHING is mutated during the outage: the reconnect's initial event necessarily carries
    // the revision already applied.
    await restart();
    await delay(1500);

    // The store gates dispatch on the content-addressed revision, so a repeated revision is not
    // a change and must wake nobody.
    expect(fires).toHaveLength(0);

    // Proof the stream and the watcher were live the whole time: a REAL change still fires once.
    await activate(engine, 'agentic', { 'agentic.lanes.vinci': { enabled: true, model_target_ref: 'changed' } }, 1);
    expect(await until(() => fires.length > 0, 15_000)).toBe(true);
    await delay(300);
    expect(fires).toEqual([{ enabled: true, model_target_ref: 'changed' }]);

    await manager.close();
  }, 40_000);
});

File: packages/go/varrpc/subscribe_sync_test.go
package varrpc

import (
	"context"
	"encoding/json"
	"net"
	"sync/atomic"
	"testing"
	"time"

	cnos "github.com/kitsyai/cnos/packages/go"
)

// SELF-SYNCHRONIZING SUBSCRIBE (round-3 follow-up), Go half.
//
// Round-3 blocker 1 ("rpc reconnect never re-pulls subscribed scopes") was fixed on the CLIENT
// side. The server still registered its commit listener only AFTER authorization resolved, so a
// commit landing between the Subscribe request arriving and that registration completing reached
// neither the stream nor the resync pull. Losing a DEACTIVATION there leaves a consumer serving
// withdrawn policy with no poller to converge — subscribe-capable sources deliberately do not
// poll.
//
// The Go client needs no change to accept the new initial event: it is an ordinary
// batch / no_head message on the existing ingest path. These tests PROVE that rather than
// assuming it, and drive the authorization-window race deterministically via the test server's
// auth gate instead of hoping to hit a sub-millisecond window.

// gatedServer starts a test server whose authorize blocks until the returned release is called.
func gatedServer(t *testing.T, requiredToken string) (*testServer, *runningServer, func()) {
	t.Helper()

	service := newTestServer()
	service.authGate = make(chan struct{})
	service.requiredToken = requiredToken
	server := serveOn(t, service, nil)
	t.Cleanup(server.stop)

	released := false
	return service, server, func() {
		if !released {
			released = true
			close(service.authGate)
		}
	}
}

// waitFor polls until the predicate holds, or fails the test.
func waitFor(t *testing.T, what string, predicate func() bool) {
	t.Helper()
	deadline := time.Now().Add(10 * time.Second)
	for time.Now().Before(deadline) {
		if predicate() {
			return
		}
		time.Sleep(10 * time.Millisecond)
	}
	t.Fatalf("timed out waiting for %s", what)
}

func TestCommitInTheAuthorizeWindowIsBufferedAndDeliveredExactlyOnce(t *testing.T) {
	service, server, release := gatedServer(t, "")

	provider := newProvider(t, server.target, nil, "")
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	received := make(chan cnos.VarBatchResult, 8)
	stop, err := provider.Subscribe(ctx, []cnos.VarScope{{Group: "agentic"}}, func(batch cnos.VarBatchResult) {
		received <- batch
	})
	if err != nil {
		t.Fatalf("subscribe: %v", err)
	}
	defer stop()

	// The server is now inside its authorization window with the subscriber queue already
	// registered. Before the fix, the listener did not exist yet and this commit was dropped.
	waitFor(t, "the server to enter the authorize window", func() bool { return service.authWindowEntered() >= 1 })

	service.activate("agentic", 4, "sha256:in-window", map[string]any{
		"agentic.lanes.vinci": map[string]any{"enabled": true, "model_target_ref": "in-window"},
	})

	select {
	case batch := <-received:
		t.Fatalf("nothing may be written while authorization is pending, got %#v", batch)
	case <-time.After(200 * time.Millisecond):
	}

	release()

	select {
	case batch := <-received:
		if batch.Status != cnos.VarPullOK || batch.Revision != "sha256:in-window" {
			t.Fatalf("the commit made in the authorize window was lost or mangled: %#v", batch)
		}
		document, _ := batch.Values["agentic.lanes.vinci"].(map[string]any)
		if document["model_target_ref"] != "in-window" {
			t.Fatalf("values: %#v", batch.Values)
		}
	case <-time.After(10 * time.Second):
		t.Fatal("a commit landing in the authorize window was never delivered")
	}

	// EXACTLY once: the flushed buffer and the initial state carry the same revision, and the
	// initial state is deduplicated against the flush rather than repeated.
	select {
	case batch := <-received:
		t.Fatalf("the same revision was delivered twice: %#v", batch)
	case <-time.After(500 * time.Millisecond):
	}
}

func TestAuthorizationFailureTerminatesTheStreamAndBuffersNothing(t *testing.T) {
	service, server, release := gatedServer(t, "good-token")

	failures := make(chan bool, 8)
	def := cnos.VarSourceDef{Transport: "rpc", URL: server.target, Auth: map[string]string{"bearer": "secret.ops.token"}}
	providerCtx := cnos.VarProviderContext{ResolveSecret: func(string) (string, error) { return "wrong-token", nil }}
	provider, err := New(def, providerCtx, WithOnError(func(_ error, terminal bool, _ []string) {
		failures <- terminal
	}))
	if err != nil {
		t.Fatalf("new provider: %v", err)
	}
	t.Cleanup(func() { _ = provider.Close() })

	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	received := make(chan cnos.VarBatchResult, 8)
	stop, err := provider.Subscribe(ctx, []cnos.VarScope{{Group: "agentic"}}, func(batch cnos.VarBatchResult) {
		received <- batch
	})
	if err != nil {
		t.Fatalf("subscribe: %v", err)
	}
	defer stop()

	waitFor(t, "the server to enter the authorize window", func() bool { return service.authWindowEntered() >= 1 })
	service.activate("agentic", 5, "sha256:refused", map[string]any{
		"agentic.lanes.vinci": map[string]any{"enabled": true, "model_target_ref": "refused"},
	})

	release()

	// Terminal, per the unchanged UNAUTHENTICATED policy — buffering does not weaken it.
	select {
	case terminal := <-failures:
		if !terminal {
			t.Fatal("an auth-rejected Subscribe must be TERMINAL")
		}
	case <-time.After(10 * time.Second):
		t.Fatal("an auth-rejected Subscribe was never reported")
	}

	// And the buffered commit was discarded, never written to a refused identity.
	select {
	case batch := <-received:
		t.Fatalf("a refused subscription must write nothing, got %#v", batch)
	case <-time.After(500 * time.Millisecond):
	}
}

func TestFreshSubscribeImmediatelyReceivesTheCurrentHead(t *testing.T) {
	service := newTestServer()
	server := serveOn(t, service, nil)
	defer server.stop()

	service.activate("agentic", 3, "sha256:already-live", map[string]any{
		"agentic.lanes.vinci": map[string]any{"enabled": true, "model_target_ref": "already-live"},
	})

	provider := newProvider(t, server.target, nil, "")
	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
	defer cancel()

	received := make(chan cnos.VarBatchResult, 8)
	stop, err := provider.Subscribe(ctx, []cnos.VarScope{{Group: "agentic"}}, func(batch cnos.VarBatchResult) {
		received <- batch
	})
	if err != nil {
		t.Fatalf("subscribe: %v", err)
	}
	defer stop()

	// NOTHING is committed after the subscribe: this can only be the initial state event, and
	// the Go client accepts it as an ordinary batch with no special handling.
	select {
	case batch := <-received:
		if batch.Status != cnos.VarPullOK || batch.Generation != 3 || batch.Revision != "sha256:already-live" {
			t.Fatalf("initial event: %#v", batch)
		}
		document, _ := batch.Values["agentic.lanes.vinci"].(map[string]any)
		if document["model_target_ref"] != "already-live" {
			t.Fatalf("values: %#v", batch.Values)
		}
	case <-time.After(10 * time.Second):
		t.Fatal("an accepted Subscribe never emitted the current head")
	}
}

func TestSubscribeToADeactivatedScopeImmediatelyReceivesNoHead(t *testing.T) {
	service := newTestServer()
	server := serveOn(t, service, nil)
	defer server.stop()

	service.activate("agentic", 1, "sha256:r1", map[string]any{"agentic.lanes.vinci": map[string]any{"enabled": true}})
	service.deactivate("agentic")

	provider := newProvider(t, server.target, nil, "")
	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
	defer cancel()

	received := make(chan cnos.VarBatchResult, 8)
	stop, err := provider.Subscribe(ctx, []cnos.VarScope{{Group: "agentic"}}, func(batch cnos.VarBatchResult) {
		received <- batch
	})
	if err != nil {
		t.Fatalf("subscribe: %v", err)
	}
	defer stop()

	select {
	case batch := <-received:
		if batch.Status != cnos.VarPullNoHead || batch.Scope != "agentic" {
			t.Fatalf("expected an immediate no_head for a deactivated scope, got %#v", batch)
		}
	case <-time.After(10 * time.Second):
		t.Fatal("a subscribe to a deactivated scope never emitted no_head")
	}
}

// streamOnlyFactory builds the rpc provider with the SDK's OnSubscriptionConnected resync seam
// REPLACED by a counter that pulls nothing. The SDK wires that seam itself
// (`variables.resyncSubscribedScopes`), so simply not setting it on our own context would not
// have neutralized anything — the reconnect pull would still have run and these tests would
// have passed no matter what the server did. The counter proves the reconnect happened while
// guaranteeing no pull was issued on its behalf.
func streamOnlyFactory(resyncs *int64) cnos.VarSourceProviderFactory {
	return cnos.VarSourceProviderFactory{
		Transport: Transport,
		Create: func(def cnos.VarSourceDef, providerCtx cnos.VarProviderContext) (cnos.VarSourceProvider, error) {
			providerCtx.OnSubscriptionConnected = func(_ []string, _ bool) {
				atomic.AddInt64(resyncs, 1)
			}
			return New(def, providerCtx)
		},
	}
}

// streamOnlyRuntime builds a full SDK runtime over an rpc source whose reconnect resync pull is
// neutralized, so any convergence observed came from the subscription stream ALONE. pollInterval
// is declared deliberately: a subscribe-capable source must IGNORE it (capability rule), so no
// poller can converge either.
func streamOnlyRuntime(t *testing.T, target string, resyncs *int64) *cnos.Runtime {
	t.Helper()

	projection := cnos.ServerProjection{
		Version:           1,
		Workspace:         "api",
		Profile:           "stage",
		ResolvedAt:        "2026-07-20T00:00:00Z",
		ConfigHash:        "hash",
		Values:            map[string]any{"value.flags.mode": "static-tier"},
		Derived:           map[string]cnos.DerivedFormula{},
		SecretRefs:        map[string]cnos.SecretReference{},
		PublicKeys:        []string{},
		RuntimeNamespaces: []string{},
		Meta:              cnos.ProjectionMeta{Workspace: "api", Profile: "stage", CnosVersion: "1.17.0"},
		VarSources:        map[string]cnos.VarSourceDef{"svc": {Transport: "rpc", URL: target, PollInterval: "20ms"}},
		Vars:              map[string]cnos.VarGroupDef{"flags": {Source: "svc", Mode: "prefetch"}},
		Schema:            map[string]cnos.VarKeyRule{"var.flags.mode": {Type: "string"}},
	}

	payload, err := json.Marshal(projection)
	if err != nil {
		t.Fatalf("marshal projection: %v", err)
	}

	runtime, err := cnos.LoadProjection(payload, cnos.Options{
		SecretHome:         t.TempDir(),
		Environment:        map[string]string{},
		VarSourceProviders: []cnos.VarSourceProviderFactory{streamOnlyFactory(resyncs)},
	})
	if err != nil {
		t.Fatalf("load projection: %v", err)
	}
	t.Cleanup(func() { _ = runtime.Close() })

	if err := runtime.StartVars(context.Background()); err != nil {
		t.Fatalf("start vars: %v", err)
	}
	return runtime
}

// restartOnSamePort brings a fresh server up on the same address. The caller has already
// stopped the previous one and made whatever mutation the client must miss.
func restartOnSamePort(t *testing.T, service *testServer, target string) *runningServer {
	t.Helper()

	listener, err := net.Listen("tcp", target)
	if err != nil {
		t.Fatalf("re-listen on %s: %v", target, err)
	}
	restarted := serveOn(t, service, listener)
	t.Cleanup(restarted.stop)
	return restarted
}

func TestReconnectConvergesFromTheStreamAloneOnAnActivation(t *testing.T) {
	service := newTestServer()
	listener, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatalf("listen: %v", err)
	}
	target := listener.Addr().String()
	server := serveOn(t, service, listener)

	service.activate("flags", 1, "sha256:before", map[string]any{"flags.mode": "before"})

	var resyncs int64
	runtime := streamOnlyRuntime(t, target, &resyncs)
	if value, _, _ := runtime.Var("flags.mode"); value != "before" {
		t.Fatalf("expected the pre-outage head, got %v", value)
	}

	time.Sleep(300 * time.Millisecond) // let the subscription establish

	// EXACTLY ONE mutation, entirely while the client is disconnected.
	server.stop()
	time.Sleep(100 * time.Millisecond)
	service.activate("flags", 2, "sha256:during-outage", map[string]any{"flags.mode": "during-outage"})

	restartOnSamePort(t, service, target)

	deadline := time.Now().Add(20 * time.Second)
	for time.Now().Before(deadline) {
		if value, _, _ := runtime.Var("flags.mode"); value == "during-outage" {
			if atomic.LoadInt64(&resyncs) == 0 {
				t.Fatal("the subscription never reconnected — the test proved nothing")
			}
			return // converged with no resync pull and no poller anywhere
		}
		time.Sleep(25 * time.Millisecond)
	}
	t.Fatal("an activation made during the outage never converged from the stream alone")
}

func TestReconnectConvergesFromTheStreamAloneOnADeactivation(t *testing.T) {
	service := newTestServer()
	listener, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatalf("listen: %v", err)
	}
	target := listener.Addr().String()
	server := serveOn(t, service, listener)

	service.activate("flags", 1, "sha256:before", map[string]any{"flags.mode": "runtime-tier"})

	var resyncs int64
	runtime := streamOnlyRuntime(t, target, &resyncs)
	if value, _, _ := runtime.Var("flags.mode"); value != "runtime-tier" {
		t.Fatalf("expected the runtime tier, got %v", value)
	}

	time.Sleep(300 * time.Millisecond)

	// The worst case, and the one with NO other path to convergence: a withdrawal the client
	// never hears about, on a source that runs no poller. EXACTLY ONE mutation.
	server.stop()
	time.Sleep(100 * time.Millisecond)
	service.deactivate("flags")

	restartOnSamePort(t, service, target)

	deadline := time.Now().Add(20 * time.Second)
	for time.Now().Before(deadline) {
		if value, ok, _ := runtime.Var("flags.mode"); ok && value == "static-tier" {
			snapshot, _ := runtime.VarSnapshot("flags.mode")
			if snapshot.Source != cnos.VarSourceStatic {
				t.Fatalf("snapshot source after the missed deactivation: %#v", snapshot)
			}
			if atomic.LoadInt64(&resyncs) == 0 {
				t.Fatal("the subscription never reconnected — the test proved nothing")
			}
			return
		}
		time.Sleep(25 * time.Millisecond)
	}
	t.Fatal("a deactivation made during the outage never converged from the stream alone")
}

func TestInitialEventRepeatingAKnownRevisionDoesNotFireWatchers(t *testing.T) {
	service := newTestServer()
	listener, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatalf("listen: %v", err)
	}
	target := listener.Addr().String()
	server := serveOn(t, service, listener)

	service.activate("flags", 1, "sha256:stable", map[string]any{"flags.mode": "stable"})

	var resyncs int64
	runtime := streamOnlyRuntime(t, target, &resyncs)
	if value, _, _ := runtime.Var("flags.mode"); value != "stable" {
		t.Fatalf("expected the head, got %v", value)
	}

	fires := make(chan any, 16)
	unwatch := runtime.Watch("var.flags.mode", func(next, _ cnos.Snapshot) { fires <- next.Value })
	defer unwatch()

	time.Sleep(300 * time.Millisecond)

	// NOTHING is mutated across the restart, so the reconnect's initial event necessarily
	// carries the revision already applied.
	server.stop()
	time.Sleep(100 * time.Millisecond)
	restartOnSamePort(t, service, target)
	time.Sleep(1500 * time.Millisecond)

	// The store gates dispatch on the content-addressed revision, so a repeated revision is not
	// a change and must wake nobody.
	select {
	case value := <-fires:
		t.Fatalf("an initial event repeating a known revision fired a watcher with %#v", value)
	default:
	}

	// Proof the stream and the watcher were live the whole time: a REAL change still fires.
	service.activate("flags", 2, "sha256:changed", map[string]any{"flags.mode": "changed"})

	select {
	case value := <-fires:
		if value != "changed" {
			t.Fatalf("watcher fired with %#v", value)
		}
	case <-time.After(15 * time.Second):
		t.Fatal("a real change never reached the watcher — the stream was not live")
	}
}
